### PR TITLE
chore(wallet): add GetEthTokenInfo and refactor

### PR DESCRIPTION
**Pull request by onyb**
Removes dependency on Etherscan Pro API (costs us $200/mo), and allows us to pre-fill fields like name, symbol, and decimals for _any_ token on _any_ EVM chain.

Resolves https://github.com/brave/brave-browser/issues/33314

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

- Adding ERC20 tokens and NFTs should work as expected.
- Editing an existing NFT should work as expected.
- (new) Adding an ERC20 token on _any_ EVM chain should pre-fill the name, symbol, and decimals fields.

### CODE REVIEW

File: components/brave_wallet/browser/asset_ratio_response_parser.cc [hunk 1]
```diff
@@ -173,39 +171,6 @@ bool ParseAssetPriceHistory(const base::Value& json_value,
   return true;
 }
 
-mojom::BlockchainTokenPtr ParseTokenInfo(const base::Value& json_value,
-                                         const std::string& chain_id,
-                                         mojom::CoinType coin) {
-  auto token_info = api::asset_ratio::TokenInfo::FromValue(json_value);
```
Comment (supermassive) [inline components/brave_wallet/browser/asset_ratio_response_parser.cc:179]: Let's also drop `TokenInfo` and `TokenInfoResult` from asset_ratio.idl

File: components/brave_wallet/common/brave_wallet.mojom [hunk 2]
```diff
@@ -1335,6 +1328,34 @@ interface JsonRpcService {
                       string chain_id) => (string decimals,
                                            ProviderError error,
                                            string error_message);
+
+  // Fetches name of an ERC20 or ERC721 token.
+  GetEthTokenName(string contract_address,
```
Comment (supermassive) [inline components/brave_wallet/common/brave_wallet.mojom:1333]: Is it possible not to expose GetEthTokenName and GetEthTokenDecimals + GetEthTokenSymbol above in mojom, but reuse GetEthTokenInfo for
https://github.com/brave/brave-core/blob/c65fcbaf0f5d2f8527e75a76cbfc3bfd06aea020/components/brave_wallet_ui/components/extension/swap/index.tsx#L141-L181

File: components/brave_wallet_ui/common/hooks/use-pending-transaction.ts [hunk 3]
```diff
@@ -604,20 +598,25 @@ export const usePendingTransactions = () => {
...
+  const { tokenInfo: erc20ApproveTokenInfo } = useGetTokenInfo(
+    transactionDetails?.recipient &&
+      txCoinType &&
       transactionInfo?.txType === BraveWallet.TransactionType.ERC20Approve
-    ) {
-      onFindTokenInfoByContractAddress(transactionDetails.recipient)
-    }
-  }, [transactionDetails?.recipient, transactionInfo?.txType])
+      ? {
+          contractOrMintAddress: transactionDetails.recipient,
```
Comment (supermassive) [inline components/brave_wallet_ui/common/hooks/use-pending-transaction.ts:606]: Should it have Solanas term `Mint` in `contractOrMintAddress` if it is EVM only?

File: components/brave_wallet_ui/components/shared/add-custom-token-form/add-custom-token-form.tsx [hunk 4]
```diff
@@ -262,17 +241,23 @@ export const AddCustomTokenForm = (props: Props) => {
...
-  const tokenNameError = tokenName === ''
-  const tokenSymbolError = tokenSymbol === ''
-  const tokenDecimalsError = tokenDecimals === ''
-  const customAssetsNetworkError = !customAssetsNetwork?.chainId
+  const isDecimalDisabled =
+    isTokenInfoLoading ||
+    tokenInfo?.isErc721 ||
+    tokenInfo?.isErc1155 ||
+    tokenInfo?.isNft ||
+    false
```
[the remark this answers is not in the document]
Comment (onyb) [reply components/brave_wallet_ui/components/shared/add-custom-token-form/add-custom-token-form.tsx]: c0b57fd

File: components/brave_wallet/browser/asset_ratio_response_parser.cc (hunk 1 above)
Comment (onyb) [reply components/brave_wallet/browser/asset_ratio_response_parser.cc:179]: ed535e0

File: components/brave_wallet_ui/common/hooks/use-pending-transaction.ts (hunk 3 above)
Comment (onyb) [reply components/brave_wallet_ui/common/hooks/use-pending-transaction.ts]: Fixed here: 9549c59

I do intend to add support for Solana as well, but will change back to `contractOrMintAddress` once it is done.

File: components/brave_wallet/common/brave_wallet.mojom (hunk 2 above)
Comment (onyb) [reply components/brave_wallet/common/brave_wallet.mojom]: Fixed here: 17a0e1f

Also cc: @StephenHeaps, since he may be using those removed methods.

**Review by supermassive — APPROVED** [review]
wallet core lgtm

**Review by simoarpe — APPROVED** [review]
👍 `AddAssetActivity` will be completely removed soon, so part of the class could be safely commented.

**Pull request merged**